metals/src/main/scala/scala/meta/internal/metals/codeactions/BracelessBracefulSwitchCodeAction.scala

```diff
@@ -0,0 +1,558 @@
+package scala.meta.internal.metals.codeactions
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.Case
+import scala.meta.Defn
+import scala.meta.Pkg
+import scala.meta.Template
+import scala.meta.Term
+import scala.meta.Tree
+import scala.meta.inputs.Position
+import scala.meta.internal.metals.Buffers
+import scala.meta.internal.metals.CodeAction
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.parsing.Trees
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.CancelToken
+
+import org.eclipse.lsp4j.CodeActionParams
+import org.eclipse.lsp4j.TextEdit
+import org.eclipse.{lsp4j => l}
+
+class BracelessBracefulSwitchCodeAction(
+    trees: Trees,
+    buffers: Buffers
+) extends CodeAction {
+  override def kind: String = l.CodeActionKind.RefactorRewrite
+
+  override def contribute(params: CodeActionParams, token: CancelToken)(implicit
+      ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = Future {
+    val path = params.getTextDocument().getUri().toAbsolutePath
+    val range = params.getRange()
+    val maybeTree =
+      if (range.getStart == range.getEnd)
+        trees
+          .findLastEnclosingAt[Tree](
+            path,
+            range.getStart(),
+            applyWithSingleFunction
+          )
+      else None
+
+    val result =
+      for {
+        tree <- maybeTree
+        document <- buffers.get(path)
+      } yield {
+        tree match {
+          case _: Pkg => None.toSeq
+          case classDefn: Defn.Class =>
+            createCodeActionForTemplateHolder(
+              classDefn,
+              path,
+              document,
+              classDefn.templ,
+              "class definition"
+            ).toSeq
+          case objectDefn: Defn.Object =>
+            createCodeActionForTemplateHolder(
+              objectDefn,
+              path,
+              document,
+              objectDefn.templ,
+              "object definition"
+            ).toSeq
+          case traitDefn: Defn.Trait =>
+            createCodeActionForTemplateHolder(
+              traitDefn,
+              path,
+              document,
+              traitDefn.templ,
+              "trait definition"
+            ).toSeq
+          case enumDefn: Defn.Enum =>
+            createCodeActionForTemplateHolder(
+              enumDefn,
+              path,
+              document,
+              enumDefn.templ,
+              "enum definition"
+            ).toSeq
+          case valDefn: Defn.Val =>
+            createCodeActionForPotentialBlockHolder(
+              valDefn,
+              path,
+              document,
+              valDefn.rhs,
+              "val definition"
+            ).toSeq
+          case varDefn: Defn.Var =>
```

Comment:
We should be able to detect if we are at a Block instead of having a separate case for each.

Reply:
I have tried this. It creates bugs. Because not at all times the potential block holders have a block at the position of their relevant subtree.

Reply:
If we look for Blocks themselves this would not be a problem.

Reply:
Then no code action would appear when the cursor is on the variable name.

Reply:
I think the code action for adding braces to a block or removing them, should be removed. Because in some cases such as the block of code passed as the argument of `flatMap` the braces are mandatory! Or perhaps, just in those cases when the parent is a collection method, the braces should stay.

Reply:
for flatMap it's not a problem since it's in Apply not in Block.

> Sometimes you want to add braces even when we do not have a block! for example when there is a very long multiple line chain of map/flatMap functions and such

For that case yeah, it's not 100% sure where to show the code action. And in some places there will be no Block :thinking:  Might be good to leave as is for those case, but maybe we should only show the action if the code is a bit longer than a couple of charaters? Some kind of threshold so not to always show it?

Reply:
Well, the user does not have to click on it every time! There are many other things that show up but the users just don’t click on!

Reply:
And if somebody clicks on it, the whole hassle is just an undo!

Reply:
I was thinking that we can check the number of lines threshold for deciding whether to put an `EndMarker` in place when going braceless